### DISCUSSION TITLE
Custom Scripts additional Doc Part

### DISCUSSION THREAD

**Discussion by Bloodiko** [netbox-community/netbox#7190] [Ideas]
Heya

i've tried to create some scripts, to automate imports from other services via their respective web api. 

While i may be able to write a powershell script to get the Data via API, remap the data and post it on netbox's api, this was not the goal. 

I examined the example custom script and tried to create a script which fetches Network-Container and Network Prefixes from an Infoblox API and imports them into Netbox. 
This shouldve been an fairly easy task. 

But in the end I struggled with the Imports and some Namings. 
Although they are pretty obvious if you get the scheme on which those are created, it still might be worth a small part in the custom Script Doc Part, to explain what is needed. 

For example: 

```python
#Imports - Change, depending on your needs.
from dcim.models          import Site
        ipam.models           import {something-else}
        virtualization.models import * 

from dcim.choices import SiteStatusChoices
        ipam.choices           import {something-else}
        virtualization.choices import * 
 

#Example: 
from ipam.models import Prefix
from ipam.choices import PrefixStatusChoices
```

Actually I dont even know what else i couldve Imported, those are the things i figured out with the example script. 
But those imports are definitely missing, or is there somewhere a python doc regarding those things, i did not find yet ? 

I have no idea on how to write a good Doc entry, but I just guessed this was missing. 

Pls tell me your thoughts. 

Ebiko

**Comment by candlerb** [comment]
The Python API isn't really documented, but if you look at the 'models' directories in the source code it should give you a good head start.

```
$ find netbox -name 'model*' ! -name '*test*'
netbox/netbox/models.py
netbox/virtualization/models.py
netbox/circuits/models.py
netbox/tenancy/models.py
netbox/dcim/models
netbox/users/models.py
netbox/ipam/models
netbox/extras/tests/dummy_plugin/models.py
netbox/extras/models
netbox/extras/models/models.py
```

dcim and ipam are directories, where the models are split over several files:

```
$ ls netbox/dcim/models/
__init__.py			device_component_templates.py	devices.py			racks.py
cables.py			device_components.py		power.py			sites.py
$ ls netbox/ipam/models/
__init__.py	ip.py		services.py	vlans.py	vrfs.py
```

Another good source of reference is the [scripts and reports repo](https://github.com/netbox-community/reports) which uses the Django ORM API.

I don't know exactly what you mean by "But those imports are definitely missing".  If you want help you'll have to say what version of Netbox you're running (the API has changed significantly between versions, especially for things like Choices) and show the actual complete code, how you're running it, and what error message you're getting.

Are you running your code within Netbox as a custom script, triggered from the GUI?  Or are you trying to run it standalone?  Running it standalone will certainly have problems, as you would need to set a few things up first (*).  Normally for standalone scripts you'd use [nbshell](https://netbox.readthedocs.io/en/stable/administration/netbox-shell/) for that, and then you don't need to import anything.

(*) It *is* possible to run completely standalone scripts without nbshell, but it involves several lines of Django setup, and I'd have to dig around to find the code to do that.  The problem goes away if you run CLI scripts using nbshell, or deploy '[custom scripts](https://netbox.readthedocs.io/en/stable/customization/custom-scripts/)' inside Netbox itself.

**Reply by Bloodiko** [reply]
What I meant with "But those imports are definitely missing" is the following. 

The Netbox - Custom Script Documentation only includes that you need to import extras.scripts Script, as everything is dependent on this class. 
But theres nothing written about the models i need, or similiar things like choices. 
For me that was confusing. 

It took me a while to realize that i need to import those models as well. 
Like in the custom script you published on the scripts github. 

```python
from dcim.models import DeviceRole, Platform
from django.core.exceptions import ObjectDoesNotExist
from ipam.choices import IPAddressStatusChoices
from ipam.models import IPAddress, VRF
from tenancy.models import Tenant
from virtualization.choices import VirtualMachineStatusChoices
from virtualization.models import Cluster, VirtualMachine, VMInterface
from extras.scripts import Script, StringVar, IPAddressWithMaskVar, ObjectVar, ChoiceVar, IntegerVar, TextVar
```

And in the end, i just made this Discussion, not because i needed help, i created it after i found out my mistake - and thought that it may be helpful to have this information in the docs. 

Looking into the Source Code was not helpful in my case (which i did) but i couldnt figure out what is netbox code and what is custom script source code. 
Now, with the provided commands i could figure it out too, but without definitely not. 

(Also: 
I'm running Netbox 3.0.1
I'm no loger running into errors, was working fine after i found out that i need to import the model "prefix" to fix the Class not found issue. 
I'm not that experienced in python coding yet, and i dont really have an ide for that - thus the issues.)

**Reply by candlerb** [reply]
> It took me a while to realize that i need to import those models as well.
> Like in the custom script you published on the scripts github.

I see what you mean.  I guess that a list of "most useful things to import" could be created, but it would need to be maintained to stay in sync with the source.

You do have access to dynamic documentation by using [nbshell](https://netbox.readthedocs.io/en/stable/administration/netbox-shell/):

```
$ cd /opt/netbox
$ source venv/bin/activate
$ python3 netbox/manage.py nbshell
### NetBox interactive shell (netbox3)
### Python 3.8.10 | Django 3.2.7 | NetBox 3.0.1
### lsmodels() will show available models. Use help(<model>) for more info.
```

Using `lsmodels()` gives you a list of models.  `help(<model>)`, e.g. `help(Prefix)` gives you a lot of detail about the attributes of that model.  However these models have all been pre-imported, and your own code still needs to import them.

`help(...)` does show you an import path (e.g. it shows "Help on class Prefix in module **ipam.models.ip**") - but actually that's not the preferred path to use.  There's an alias to Prefix in `ipam.models` which is more likely to remain stable.

nbshell doesn't import things like `IPAddressStatusChoices` at the top level, but you can do `from ipam.choices import IPAddressStatusChoices; help(IPAddressStatusChoices)`.

**Reply by Bloodiko** [reply]
Thanks for that, gonna mark this as answer since its definitely super helpful - especially the help part about an model. 
Hoping that this thread can help others as well on search for netbox custo script help. :D 
Edit: 
Or I dont, since i didnt create this as "Help" but rather as "Idea" thus no "Answer" option...
